synapse/logging/_terse_json.py

```diff
@@ -0,0 +1,199 @@
+    # the content we want. The original event should be considered 'frozen'.
+    for key in _event.keys():
+
+        if key in keys_to_delete:
+            continue
+
+        if isinstance(_event[key], (str, int, bool, float)) or _event[key] is None:
+            # If it's a plain type, include it as is.
+            event[key] = _event[key]
+        else:
+            # If it's not one of those basic types, write out a string
+            # representation. This should probably be a warning in development,
+            # so that we are sure we are only outputting useful data.
+            event[key] = str(_event[key])
+
+    # Add the metadata information to the event (e.g. the server_name).
+    event.update(metadata)
+
+    return event
+
+
+def TerseJSONToConsoleLogObserver(outFile: TextIO, metadata: dict) -> FileLogObserver:
+    """
+    A log observer that formats events to a flattened JSON representation.
+
+    Args:
+        outFile: The file object to write to.
+        metadata: Metadata to be added to each log object.
+    """
+
+    def formatEvent(_event: dict) -> str:
+        flattened = flatten_event(_event, metadata)
+        return dumps(flattened, ensure_ascii=False, separators=(",", ":")) + "\n"
+
+    return FileLogObserver(outFile, formatEvent)
+
+
+@attr.s
+class TerseJSONToTCPLogObserver(object):
+    """
+    An IObserver that writes JSON logs to a TCP target.
+
+    Args:
+        hs (HomeServer): The Homeserver that is being logged for.
+        host: The host of the logging target.
+        port: The logging target's port.
+        metadata: Metadata to be added to each log entry.
+    """
+
+    hs = attr.ib()
+    host = attr.ib(type=str)
+    port = attr.ib(type=int)
+    metadata = attr.ib(type=dict)
+    _buffer = attr.ib(default=attr.Factory(deque), type=deque)
+    _writer = attr.ib(default=None)
+
+    def start(self) -> None:
+        endpoint = HostnameEndpoint(self.hs.get_reactor(), self.host, self.port)
+        factory = Factory.forProtocol(Protocol)
+        self._service = ClientService(endpoint, factory)
+        self._service.startService()
+
+    def _write_loop(self) -> None:
+        """
+        Implement the write loop.
+        """
+        if self._writer:
+            return
+
+        self._writer = self._service.whenConnected()
+
+        @self._writer.addBoth
+        def _(r):
+            if isinstance(r, Failure):
+                r.printTraceback(file=sys.__stderr__)
+                self._writer = None
+                self.hs.get_reactor().callLater(1, self._write_loop)
+                return
+
+            try:
+                r.transport.write(b"\n".join(reversed(self._buffer)) + b"\n")
+                self._buffer.clear()
+            except Exception as e:
+                sys.__stderr__.write("Failed writing out logs with %s\n" % (str(e),))
+
+            self._writer = False
+            self.hs.get_reactor().callLater(1, self._write_loop)
```

Comment:
These probably should be in a finally block, in case something in here dies with an uncaught exception.

Reply:
Do we want to do this event if the buffer is empty? Should this be a loopingcall instead?